### PR TITLE
Refactor wappi animations

### PR DESCRIPTION
**Pull request by Owen-Delisle**

### CODE REVIEW

**Comment by Owen-Delisle** [comment]
[Tests.zip](https://github.com/SlantDesign/FlowTemplates/files/4738769/Tests.zip)

File: Packages/Web/lib/util.flp [hunk 1]
```diff
@@ -459,3 +459,46 @@ end
...
+    return values
+end
+
+func addTrailingKeyvalue(_ keyvalues, _ duration, _ arr, _ appendingValue)
+     if keyvalues[keyvalues.count - 1].time * 1000 < duration
+        arr = arr.appending(appendingValue)
+        return arr
+    end
+    return arr 
+end
```
Comment (jbeenie) [inline Packages/Web/lib/util.flp:496]: I think we might not need this anymore now that we're using the simulated keyvalues.

File: Packages/Web/lib/util.flp [hunk 2]
```diff
@@ -459,3 +459,46 @@ end
...
+
+func addInitialValue(_ keyvalues, _ duration, _ arr, _ appendingValue)
+    if keyvalues[0].time > 0
+        arr = [appendingValue].concatenate(arr)
+        return arr
+    end
+    return arr 
+end
```
Comment (jbeenie) [inline Packages/Web/lib/util.flp:504]: I think we might not need this anymore now that we're using the simulated keyvalues.

File: Packages/Web/lib/util.flp [hunk 3]
```diff
@@ -459,3 +459,46 @@ end
...
+
+func trackValues(property, _ keyvalues, layer, imageFolderPath, importImage, _ duration)
+    var values = []
+    for keyvalue in keyvalues
+        var value = cssValue(property: property, value: keyvalue.value, layer: layer, imageFolderPath: imageFolderPath, importImage: importImage)
+        values = values.appending(value)
+    end
+
+    values = addTrailingKeyvalue(keyvalues, duration, values, values[values.count - 1])
+    values = addInitialValue(keyvalues, duration, values, values[0])
```
Comment (jbeenie) [inline Packages/Web/lib/util.flp:485]: I think we can remove these two lines

File: Packages/Web/lib/util.flp [hunk 4]
```diff
@@ -459,3 +459,46 @@ end
 func trackName(propertyName, id)
   return "\(id)_\(propertyName.snakeCased())Track"
 end
+
+func trackOffsets(_ times, _ duration)
+    var offsets = []
+    if times[0] != 0
+        offsets = [0]
+    end
```
Comment (jbeenie) [inline Packages/Web/lib/util.flp:467]: I think we can remove this if statement now that were using simulated key values

File: Packages/Web/lib/util.flp [hunk 5]
```diff
@@ -459,3 +459,46 @@ end
...
+    for time in times
+        offsets = offsets.appending(time / duration)
+    end
+    if offsets[offsets.count - 1] != 1
+        offsets = offsets.appending(1)
+    end
```
Comment (jbeenie) [inline Packages/Web/lib/util.flp:473]: I think we can remove this if statement now that were using simulated key values

File: Packages/Web/lib/util.flp [hunk 6]
```diff
@@ -196,14 +196,14 @@ end
...
-  var timingFunctions = []
-  for keyvalue in keyvalues
-    var timingFunction = cssTimingFunction(of: keyvalue.trailingTimingFunction).inQuotes()
-    timingFunctions = timingFunctions.appending(timingFunction)
-  end
-
-  return timingFunctions
+func timingFunctions(keyvalues, _ duration)
+    var timingFunctions = []
+    for keyvalue in keyvalues.dropLast()
```
Comment (jbeenie) [inline Packages/Web/lib/util.flp:201]: lets remove the drop last here

File: Packages/Web/templates/js/track.js.flt [hunk 7]
```diff
@@ -1,19 +1,17 @@
...
-for keyvalue in keyvalues
-  var value = Web.cssValue(property: property, value: keyvalue.value, layer: layer, imageFolderPath: imageFolderPath, importImage: importImage)
-  values = values.appending(value)
-end
+
+var values = Web.trackValues(property: property, keyvalues, layer: layer, imageFolderPath: imageFolderPath, importImage: importImage, duration)
 
 // Get all the keyvalue timing functions
-var timingFunctions = Web.timingFunctions(keyvalues: keyvalues.dropLast())
+var timingFunctions = Web.timingFunctions(keyvalues: keyvalues, duration)
```
Comment (jbeenie) [inline Packages/Web/templates/js/track.js.flt:14]: lets add the drop last back over here.

File: Packages/Web/templates/js/track.js.flt [hunk 8]
```diff
@@ -1,19 +1,17 @@
...
 var times = Web.animationTimes(of: keyvalues)
 
 // Get all the keyvalue values
-var values = []
-for keyvalue in keyvalues
-  var value = Web.cssValue(property: property, value: keyvalue.value, layer: layer, imageFolderPath: imageFolderPath, importImage: importImage)
-  values = values.appending(value)
-end
+
+var values = Web.trackValues(property: property, keyvalues, layer: layer, imageFolderPath: imageFolderPath, importImage: importImage, duration)
```
Comment (jbeenie) [inline Packages/Web/templates/js/track.js.flt:11]: good idea making this a function

**Pull request merged**